pkg/cachedinstances/cloudprovider/cached_cloud_provider.go

```diff
@@ -0,0 +1,261 @@
+	statsCacheRefreshPositive uint64 // Cumulative number of positive refreshes (ie, a refresh which succeeded)
+	statsCacheRefreshNegative uint64 // Cumulative number of negative refreshes (ie, a refresh which failed and used old data)
+	statsCachePositive        uint64 // Absolute number of positive entries in cache
+	statsCacheNegative        uint64 // Absolute number of negative entries in cache
+
+	logger         logrus.FieldLogger
+	limiter        *rate.Limiter
+	cloudProvider  gostatsd.CloudProvider
+	cacheOpts      gostatsd.CacheOptions
+	ipSinkSource   chan gostatsd.IP
+	infoSinkSource chan gostatsd.InstanceInfo
+
+	// emitChan triggers a write of all the current stats when it is given a Statser
+	emitChan     chan stats.Statser
+	rw           sync.RWMutex // Protects cache
+	cache        map[gostatsd.IP]*instanceHolder
+	toLookupIPs  []gostatsd.IP
+	toReturnInfo []gostatsd.InstanceInfo
+}
+
+func (ch *CachedCloudProvider) Run(ctx context.Context) {
+	var (
+		toLookupC     chan<- gostatsd.IP
+		toLookupIP    gostatsd.IP
+		toReturnInfoC chan<- gostatsd.InstanceInfo
+		toReturnInfo  gostatsd.InstanceInfo
+		wg            wait.Group
+	)
+	// this goroutine needs to populate/update the cache so an intermediate InstanceInfo channel is used below that allows
+	// to intercept, update the cache and then push the information through to the cache consumer.
+	ownInfoSource := make(chan gostatsd.InstanceInfo)
+	ld := cloudProviderLookupDispatcher{
+		logger:        ch.logger,
+		limiter:       ch.limiter,
+		cloudProvider: ch.cloudProvider,
+		ipSource:      ch.ipSinkSource, // our sink is their source
+		infoSink:      ownInfoSource,   // their sink is our source
+	}
+
+	defer wg.Wait() // Wait for cloudProviderLookupDispatcher to stop
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel() // Tell cloudProviderLookupDispatcher to stop
+
+	wg.StartWithContext(ctx, ld.run)
+
+	refreshTicker := time.NewTicker(ch.cacheOpts.CacheRefreshPeriod)
+	defer refreshTicker.Stop()
+	// No locking for ch.cache READ access required - this goroutine owns the object and only it mutates it.
+	// So reads from the same goroutine are always safe (no concurrent mutations).
+	// When we mutate the cache, we hold the exclusive (write) lock to avoid concurrent reads.
+	// When we read from the cache from other goroutines in the Peek() method, we obtain the read lock.
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case toLookupC <- toLookupIP:
+			toLookupIP = gostatsd.UnknownIP // enable GC
+			toLookupC = nil                 // ip has been sent; if there is nothing to send, the case is disabled
+		case toReturnInfoC <- toReturnInfo:
+			toReturnInfo = gostatsd.InstanceInfo{} // enable GC
+			toReturnInfoC = nil                    // info has been sent; if there is nothing to send, the case is disabled
+		case info := <-ownInfoSource:
+			ch.handleInstanceInfo(info)
+		case t := <-refreshTicker.C:
+			ch.doRefresh(t)
```

Comment:
I would suggest skipping refreshing if we haven't managed to clear the previous refresh yet. I could see a situation where we get in to a bad state (due to cpu or network, not necessarily a bug), where we're refreshing faster than we can process, causing either unbound growth in `ch.toLookupIPs` or having something to write before we're ready if the below suggestion is done.

I'm not so sure if the same pattern is useful in `toReturnInfo`, it complicates the interface a bunch.

tl;dr backpressure on `ch.toLookupIPs`

Reply:
I agree and I thought about that. But I think it should be done separately. This PR does not make it better or worse, right? Also, don't want to make it even bigger.

Reply:
Works for me 👍